### PR TITLE
fix: show navigation button on hover in carousel content area, carousel component css moved from ngx

### PR DESCRIPTION
**Pull request by DeepakSap14**
## Related Issue
Related to Issue SAP/fundamental-ngx#5025

## Description
Moving css present for carousel component in Ngx to styles.
navigation button in content area will be visible on hover on carousel.

#### Please check whether the PR fulfills the following requirements

1. The output matches the design specs
- [x] All values are in `rem`
- [x] Text elements follow the truncation rules
- [x] hover state of the element follow design spec
- [x] focus state of the element follow design spec
- [x] active state of the element follow design spec
- [x] selected state of the element follow design spec
- [x] selected hover state of the element follow design spec
- [x] pressed state of the element follow design spec
- [x] Responsiveness rules - the component has modifier classes for all breakpoints
- [x] Includes Compact/Cosy/Tablet design
- [x] RTL support
2. The code follows fundamental-styles code standards and style
- [x] only one top level `fd-*` class is used in the file
- [x] BEM naming convention is used
- [x] Mixins are used for repeatable code (`fd-rtl`, `fd-ellipsis`, `fd-flex`, `fd-selected`, `fd-focus`, ect.)
- [x] A11y support - keyboard support, screenreader support, proper ARIA attributes, etc.
- [x] `fd-reset()` mixin is applied to all elements
- [x] Variables are used, if some value is used more than twice.
- [x] Checked if current components can be reused, instead of having new code.
3. Testing
- [x] tested Storybook examples with "CSS Resources" `normalize` option 
- [x] tested Storybook examples with "CSS Resources" `unnormalize` option 
- [x] Verified all styles in IE11
- [x] Updated tests
- [x] last commit message should have `[ci visual]` so it can trigger chromatic visual regression (e.g. `test: run chromatic visual regression [ci visual]`)
4. Documentation
- [x] Storybook documentation has been created/updated
- [NA] Breaking Changes [wiki](https://github.com/SAP/fundamental-styles/wiki/Breaking-Changes) has been updated in case of breaking changes.

### CODE REVIEW

File: src/carousel.scss [hunk 1]
```diff
@@ -172,7 +185,13 @@ $button: #{$fd-namespace}-button;
 
   &--no-navigation {
     .#{$block}__button.#{$button} {
-      display: none;
+      display: none !important;
```
Comment (aberikashvili) [inline src/carousel.scss:188]: Is important really necessary here? can we use specificity instead of `!important` here?

You can see this article why using `!important` is bad practise https://uxengineer.com/css-specificity-avoid-important-css/

**Review by aberikashvili — COMMENTED** [review]
Please see my comment regarding `!important`

**Comment by DeepakSap14** [comment]
> Please see my comment regarding `!important`

using specificity instead of `!important` now.

File: src/carousel.scss [hunk 2]
```diff
@@ -21,6 +21,14 @@ $button: #{$fd-namespace}-button;
   @include fd-reset();
   @include fd-flex(column);
 
+  @include fd-hover() {
+    .#{$block}__content {
+      .#{$block}__button.#{$button} {
```
Comment (InnaAtanasova) [inline src/carousel.scss:26]: Do you need `.#{$button}`? You can avoid this by adding more specificity.

File: src/carousel.scss [hunk 3]
```diff
@@ -174,5 +187,19 @@ $button: #{$fd-namespace}-button;
     .#{$block}__button.#{$button} {
       display: none;
     }
+
+    @include fd-hover() {
+      .#{$block}__content {
+        .#{$block}__button.#{$button} {
```
Comment (InnaAtanasova) [inline src/carousel.scss:193]: same comment as above

File: stories/carousel/carousel.stories.js [hunk 4]
```diff
@@ -477,3 +476,100 @@ carouselNoNavigation.parameters = {
...
     }
 };
+
+export const horizontalCarousel = () => `<div style="display: flex; flex-direction: column; align-items: center; background: #CCD1D1;">
+    <h4>Items in horizontal direction</h4>
+    <div
+        class="fd-carousel"
+        data-ride="carousel"
+        style="margin-bottom: 3rem; max-width: 60rem;">
+        <div class="fd-carousel__content fd-carousel--horizontal" style="text-align: center; padding: 1rem; min-height: 15.5rem;">
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:486]: wrong BEM

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

File: src/carousel.scss (hunk 2 above)
Comment (DeepakSap14) [reply src/carousel.scss]: `.#{$button}` not needed. It works fine with only `.#{$block}__button`.
corrected.

File: src/carousel.scss (hunk 3 above)
Comment (DeepakSap14) [reply src/carousel.scss]: corrected, same as above.

File: stories/carousel/carousel.stories.js (hunk 4 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: using __horizontal now.

File: stories/carousel/carousel.stories.js [hunk 5]
```diff
@@ -477,3 +476,100 @@ carouselNoNavigation.parameters = {
...
     }
 };
+
+export const horizontalCarousel = () => `<div style="display: flex; flex-direction: column; align-items: center; background: #CCD1D1;">
+    <h4>Items in horizontal direction</h4>
+    <div
+        class="fd-carousel"
+        data-ride="carousel"
+        style="margin-bottom: 3rem; max-width: 60rem;">
+        <div class="fd-carousel__content fd-carousel__horizontal" style="text-align: center; padding: 1rem; min-height: 15.5rem;">
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:486]: It's wrong again

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]

File: stories/carousel/carousel.stories.js (hunk 5 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: have moved it inside __content. so the class name now is __content--horizontal.
have created class for --error-message, as it was getting broken after above change.

**Review by aberikashvili — APPROVED** [review]
LGTM

File: stories/carousel/carousel.stories.js [hunk 6]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
...
+                    role="img"
+                    aria-label="Carousel Image 5">
+                </div>
+            </div>
+
+            <div class="fd-carousel__item fd-carousel__item--active">
+                <div
+                    style="
+                        width:100%;
+                        background-image: url(assets/images/backgrounds/Godafoss_waterfall_in_northern_Iceland.jpg);
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:503]: Is this image free from copyrights?

File: stories/carousel/carousel.stories.js [hunk 7]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
...
+        Displaying item 2 of 3
+    </div>
+</div>
+`;
+
+horizontalCarousel.storyName = 'Items in horizontal direction';
+horizontalCarousel.parameters = {
+    docs: {
+        iframeHeight: 900,
+        storyDescription: `Carousel will have items which will spread horizontally and will be visible on navigation. For translation to work, all items should be in DOM.
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:560]: `translation` ?

File: stories/carousel/carousel.stories.js [hunk 8]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
...
+    }
+};
+
+export const error = () => `<div style="display: flex; flex-direction: column; align-items: center; background: #CCD1D1;">
+    <h4>Error in loading items</h4>
+    <div
+        class="fd-carousel fd-carousel--no-navigation"
+        data-ride="carousel"
+        style="margin-bottom: 3rem; max-width: 60rem;">
+        <div class="fd-carousel__content fd-carousel__content--error-message" style="text-align: center; padding: 1rem; min-height: 15.5rem;">
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:571]: Why do you have a new class `fd-carousel__content--error-message` AND custom css applied on the same element?

File: stories/carousel/carousel.stories.js [hunk 9]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
...
+            <div class="docs-column-flex">
+                <div class="fddocs-horizontal-center">
+                    <i class="sap-icon--document"></i>
+                </div>
+                Items could not be loaded
+            </div>
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:577]: This part will not be consumed in ngx and people will have to add custom css there

File: stories/carousel/carousel.stories.js [hunk 10]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
 `
     }
 };
+
+export const horizontalCarousel = () => `<div style="display: flex; flex-direction: column; align-items: center; background: #CCD1D1;">
+    <h4>Items in horizontal direction</h4>
+    <div
+        class="fd-carousel"
+        data-ride="carousel"
+        style="margin-bottom: 3rem; max-width: 60rem;">
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:485]: Shouldn't have inline styling

File: stories/carousel/carousel.stories.js [hunk 11]
```diff
@@ -477,3 +476,133 @@ carouselNoNavigation.parameters = {
...
+        <div class="fd-carousel__content fd-carousel__content--horizontal" style="text-align: center; padding: 1rem; min-height: 15.5rem;">
```
Comment (InnaAtanasova) [inline stories/carousel/carousel.stories.js:486]: Shouldn't have inline styling

File: stories/carousel/carousel.stories.js (hunk 6 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js:503]: This image is already getting used in previous example of carousel.
https://sap.github.io/fundamental-styles/?path=/docs/components-carousel--carousel-bottom.
Not sure on copy rights, but we have few more images in `assets/images/backgrounds/` folder.

**Review by InnaAtanasova — CHANGES_REQUESTED** [review]
Reconsider the usage of inline styling on the elements that have fundamental classes

File: stories/carousel/carousel.stories.js (hunk 7 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: corrected the sentence.
For translate animation effect to work, all items should be in DOM.

File: stories/carousel/carousel.stories.js (hunk 8 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: copy paste issue from above example.
removed inline styling.

File: stories/carousel/carousel.stories.js (hunk 10 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: removed inline styling.

File: stories/carousel/carousel.stories.js (hunk 11 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: removed inline styling.

File: stories/carousel/carousel.stories.js (hunk 9 above)
Comment (DeepakSap14) [reply stories/carousel/carousel.stories.js]: It was decided during tech spec review that Error message will be coming in Ngx as <ng-content>.
so only placing of error message is take care.

formatting of error message like having Icon and Error message vertically and icon horizontally centred will have to be taken care by application.

Do you want to have that styling present here?
if so, the during application development, those classes will be exposed.

let me know your thought and correct me.

Comment (InnaAtanasova) [reply stories/carousel/carousel.stories.js]: Hi @DeepakSap14 the specs clearly mention that the Empty State (when no pages are loaded) should use the Message Page component. It's present in both fund-styles and fund-ngx. You don't have to use custom css, just use the component.
Screen Shot 2021-06-16 at 8 52 44 AM

**Comment by InnaAtanasova** [comment]
Screen Shot 2021-06-16 at 8 49 20 AM
Screen Shot 2021-06-16 at 8 49 02 AM

I think the problem with using inline styling and/or docs styling is still not clear. This styling is NOT consumed by the implementation libraries. You just make the component looks good on the documentation page, but when imported in fund-ngx, for example, you will need again custom css. 
Here's an example:
Screen Shot 2021-06-16 at 8 49 02 AM
Screen Shot 2021-06-16 at 8 49 20 AM

When you turn of the styling from the docs, the component doesn't look like as should.

**Comment by DeepakSap14** [comment]
> Screen Shot 2021-06-16 at 8 49 20 AM
> 
> Screen Shot 2021-06-16 at 8 49 02 AM
> 
> I think the problem with using inline styling and/or docs styling is still not clear. This styling is NOT consumed by the implementation libraries. You just make the component looks good on the documentation page, but when imported in fund-ngx, for example, you will need again custom css.
> 
> When you turn off the styling from the docs, the component doesn't look like as should.

@InnaAtanasova 
1. when we create carousel in Ngx, we provide width for carousel; which is width taken by single carousel item.
2. In case of horizontal carousel example, i have shown 3 carousel items; to show horizontal arrangement of items.
    ideally, only 1 item will be visible and carousel width will be equal to width taken by this carousel item, taking 100% width, as expected.
3. I am attaching a video here, where i am creating carousel of 20rem width. i am setting carousel item minimum-width to be 20rem to fill the carousel. and when only 1 item is visible, carousel looks fine.
4. let me know, if this implementation is not correct.
    

https://user-images.githubusercontent.com/57661487/122268866-374ef780-cefa-11eb-8324-d09e7b4211fe.mp4

**Comment by InnaAtanasova** [comment]
> > Screen Shot 2021-06-16 at 8 49 20 AM
> > Screen Shot 2021-06-16 at 8 49 02 AM
> > I think the problem with using inline styling and/or docs styling is still not clear. This styling is NOT consumed by the implementation libraries. You just make the component looks good on the documentation page, but when imported in fund-ngx, for example, you will need again custom css.
> > When you turn off the styling from the docs, the component doesn't look like as should.
> 
> @InnaAtanasova
> 
> 1. when we create carousel in Ngx, we provide width for carousel; which is width taken by single carousel item.
> 2. In case of horizontal carousel example, i have shown 3 carousel items; to show horizontal arrangement of items.
>    ideally, only 1 item will be visible and carousel width will be equal to width taken by this carousel item, taking 100% width, as expected.
> 3. I am attaching a video here, where i am creating carousel of 20rem width. i am setting carousel item minimum-width to be 20rem to fill the carousel. and when only 1 item is visible, carousel looks fine.
> 4. let me know, if this implementation is not correct.
> 
>  2021-06-16_23-21-59.mp4

@Hi @DeepakSap14 the building blocks of Carousel are made in a dynamic way to fit any content, aka the width of the elements is `100%`. Here's what I mean:
In the image bellow you will see the changes I made on your example:
1. I removed the custom css that you added
2. I added width to the images. You had before 100% but 100% of what? The images will shrink to fit the available content. When you provide width of the inner elements, the images in this case, the Carousel elements will grow to fit them. 
3. Instead of adding margin bottom you can just provide height for the example container.

What you will achieve with this:
- less code in the documentation - you don't need an extra div and custom css
- you will not confuse the user when they copy the example and paste it in their code about the need if the custom css
- show the user how to manage the width of the content inside the Carousel 

I hope this makes sense.

Screen Shot 2021-06-16 at 4 26 35 PM

**Comment by InnaAtanasova** [comment]
None of these css classes are needed in the documentation:
Screen Shot 2021-06-16 at 4 54 39 PM

**Review by droshev — DISMISSED** [review]
@DeepakSap14 are you sure you are fixing the whole #5052? I doubt that

File: .storybook/custom/custom.scss [hunk 12]
```diff
@@ -700,3 +700,20 @@ body.sb-show-main {
 .fddocs-list-height {
   height: 11rem;
 }
+
```
Comment (droshev) [inline .storybook/custom/custom.scss:703]: why is this added?

**Comment by DeepakSap14** [comment]
> @DeepakSap14 are you sure you are fixing the whole #5052? I doubt that

corrected it.

**Comment by DeepakSap14** [comment]
> None of these css classes are needed in the documentation:
> Screen Shot 2021-06-16 at 4 54 39 PM

@InnaAtanasova 
custom classes are removed.
for showing Error message, using message page component.
examples created correctly as suggested by you.

please see if it looks fine.

**Review by InnaAtanasova — APPROVED** [review]
